client/src/testing.ts

```diff
@@ -0,0 +1,341 @@
+// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
+
+import {
+  type TestData,
+  type TestIdentifier,
+  type TestMessage,
+  testModule,
+  testModuleDelete,
+  testRun,
+  testRunCancel,
+  testRunProgress,
+} from "./lsp_extensions";
+import type { DenoExtensionContext } from "./types";
+import { assert } from "./util";
+
+import * as vscode from "vscode";
+import { MarkupKind } from "vscode-languageclient/node";
+import type {
+  ClientCapabilities,
+  DocumentSelector,
+  MarkupContent,
+  ServerCapabilities,
+  StaticFeature,
+} from "vscode-languageclient/node";
+
+/** Indicate to the language server that the client supports the Testing API */
+export class TestingFeature implements StaticFeature {
+  #enabled = false;
+
+  get enabled() {
+    return this.#enabled;
+  }
+
+  fillClientCapabilities(capabilities: ClientCapabilities): void {
+    const experimental: { testingApi?: boolean } = capabilities.experimental ??
+      {};
+    experimental.testingApi = true;
+    capabilities.experimental = experimental;
+  }
+
+  initialize(
+    capabilities: ServerCapabilities<{ testingApi?: boolean }>,
+    _documentSelector: DocumentSelector | undefined,
+  ): void {
+    this.#enabled = capabilities.experimental?.testingApi ?? false;
+  }
+
+  dispose(): void {}
+}
+
+function asStringOrMarkdown(
+  content: MarkupContent,
+): string | vscode.MarkdownString {
+  switch (content.kind) {
+    case MarkupKind.Markdown:
+      return new vscode.MarkdownString(content.value);
+    case MarkupKind.PlainText:
+      return content.value;
+  }
+}
+
+/** Resolve the test that a test item belongs to. If the test item is a test,
+ * the result will be itself, but if the test item is a step, the result will
+ * be the owning test. */
+function getTest(testItem: vscode.TestItem): vscode.TestItem {
+  let item = testItem;
+  while (item.parent && item.parent.parent) {
```

Comment:
Is ` && item.parent.parent` there because the root is the file?

Reply:
Correct, the test item the represents a module that contains a test will have `parent === undefined`.